operators/src/pro/cache/tile_cache.rs

```diff
@@ -29,16 +30,425 @@ pub struct TileCacheBackend {
+            0
+        } else {
+            self.byte_size_total - self.byte_size_used
+        }
+    }
+
+    #[inline]
+    pub fn size_used_fraction(&self) -> f64 {
+        self.byte_size_used as f64 / self.byte_size_total as f64
+    }
+
+    #[inline]
+    pub fn can_fit_bytes(&self, bytes: usize) -> bool {
+        self.byte_size_free() >= bytes
+    }
+
+    #[inline]
+    pub fn can_fit_element_bytes<T: ByteSize>(&self, element: &T) -> bool {
+        self.can_fit_bytes(element.byte_size())
+    }
+
+    #[inline]
+    pub fn is_overfull(&self) -> bool {
+        self.byte_size_used > self.byte_size_total
+    }
+
+    #[inline]
+    fn add_bytes(&mut self, bytes: usize) {
+        debug_assert!(
+            self.can_fit_bytes(bytes),
+            "adding too many bytes {} for free capacity {}",
+            bytes,
+            self.byte_size_free()
+        );
+        self.byte_size_used += bytes;
+    }
+
+    #[inline]
+    #[allow(dead_code)] // used for tests
+    fn add_element_bytes<T: ByteSize>(&mut self, element: &T) {
+        self.add_bytes(element.byte_size());
+    }
+
+    #[inline]
+    fn remove_bytes(&mut self, bytes: usize) {
+        debug_assert!(
+            self.byte_size_used >= bytes,
+            "removing more bytes {} than used {}",
+            bytes,
+            self.byte_size_used
+        );
+        self.byte_size_used -= bytes;
+    }
+
+    #[inline]
+    fn remove_element_bytes<T: ByteSize>(&mut self, element: &T) {
+        self.remove_bytes(element.byte_size());
+    }
+
+    #[inline]
+    pub fn try_add_bytes(&mut self, bytes: usize) -> Result<(), CacheError> {
+        ensure!(
+            self.can_fit_bytes(bytes),
+            crate::pro::cache::error::NotEnoughSpaceInCache
+        );
+        self.add_bytes(bytes);
+        Ok(())
+    }
+
+    #[inline]
+    pub fn try_add_element_bytes<T: ByteSize>(&mut self, element: &T) -> Result<(), CacheError> {
+        let bytes = element.byte_size();
+        self.try_add_bytes(bytes)
+    }
+
+    #[inline]
+    pub fn try_remove_bytes(&mut self, bytes: usize) -> Result<(), CacheError> {
+        ensure!(
+            self.byte_size_used >= bytes,
+            crate::pro::cache::error::NegativeSizeOfCache
+        );
+        self.remove_bytes(bytes);
+        Ok(())
+    }
+
+    #[inline]
+    pub fn try_remove_element_bytes<T: ByteSize>(&mut self, element: &T) -> Result<(), CacheError> {
+        let bytes = element.byte_size();
+        self.try_remove_bytes(bytes)
+    }
+
+    #[inline]
+    fn add_bytes_allow_overflow(&mut self, bytes: usize) {
+        self.byte_size_used += bytes;
+        if self.byte_size_used > self.byte_size_total {
+            trace!(
+                "overflowing cache size by {} bytes, total size: {}, added bytes: {}",
+                self.byte_size_used - self.byte_size_total,
+                self.byte_size_total,
+                bytes
+            );
+        }
+    }
+}
+
+type CacheSizeLruRef<'a> = (
```

Comment:
why do you need these helper types?
Maybe make a struct out of it instead so it's more apparent what each element is?

Same for the other helper types

Reply:
clippy wanted that type

Reply:
i do in 379 and 440

Reply:
> clippy wanted that type

why though? Why doesn't it let you access two fields of you struct? What was the error message?

Reply:
[0739e8b](https://github.com/geo-engine/geoengine/pull/819/commits/0739e8b3e781ca3c6444c168e43f815b21ff6d9d)